PYIC-7309: Add accessibility check for template pages

MikeCollingwood opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

- Add accessibility check for template pages
- Make browser tests run in parallel for performance

### Why did it change

- To build out our core-front tests to include accessibility testing

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-7309](https://govukverify.atlassian.net/browse/PYIC-7309)

[PYIC-7309]: https://govukverify.atlassian.net/browse/PYIC-7309?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

MikeCollingwood committed:
PYIC-7309:
- Add accessibility check for template pages
- Make browser tests run in parallel for performance

MikeCollingwood committed:
PYIC-7309:
- Test without synced toggle

MikeCollingwood committed:
PYIC-7309: Sync language toggle

blakeyp commented on browser-tests/functional-tests/accessibility.spec.js:
```diff
@@ -0,0 +1,32 @@
+  iteratePagesAndContexts(
+    (pageName, context, language, url) => {
+      test(`Accessibility check for ${pageName}, context ${context} and language ${language}`, async ({ page }) => {
+        await page.goto(url);
```
Some of our pages have a details component that we should probably open before running the scan if we can. See https://playwright.dev/docs/accessibility-testing#configuring-axe-to-scan-a-specific-part-of-a-page
(is it possible to just try opening a details component if it's there?)

MikeCollingwood replied:
I will look at implementing this but maybe just in a test for a single page, like below for the error state? Testing for other pages is unlikely to increase coverage meaningfully and this way the tests run quicker

blakeyp replied:
Fine to leave it if it's a faff

blakeyp commented on browser-tests/functional-tests/accessibility.spec.js:
```diff
@@ -0,0 +1,32 @@
+  (["radio", "checkbox"]).forEach((errorState) => {
+    ["en", "cy"].forEach((language) => {
+      test(`Accessibility check for ${errorState} error state in ${language}`, async ({ page }) => {
+        await page.goto(`http://localhost:4601/dev/template/confirm-your-details/${language}?errorState=${errorState}`);
```
Probably best to get the domain for this from config or at least define it in one place somewhere?

blakeyp replied:
Also what does setting the error state actually do on this specific template 🤔

MikeCollingwood replied:
I thought that myself and commented the same thing on the Snapshot tests PR. This was the response: [thread](https://github.com/govuk-one-login/ipv-core-front/pull/1574#discussion_r1758825390), so I've gone to keep it the same here. If you think it's still worth it I can see what I can do.

On your second point, I wrote a bit on it in the [README for pages on creating pages](https://github.com/govuk-one-login/ipv-core-front/blob/main/src/views/README.md#:~:text=pageErrorState,/%20radiobox). Our error handling has redundancy - we have _two_ objects which lead to error states, but this is the only one where can specify whether it is the radio or checkbox which is failing. It causes either the radio buttons or the checkboxes to show validation errors

blakeyp replied:
Can't we pull it from the env vars defined in the docker compose file?

MikeCollingwood committed:
PYIC-7309: Add Accessibility check for open gov uk details

blakeyp commented on browser-tests/functional-tests/accessibility.spec.js:
```diff
@@ -0,0 +1,40 @@
+  ["en", "cy"].forEach((language) => {
+    test(`Accessibility check for open details in ${language}`, async ({ page }) => {
+      await page.goto(`http://localhost:4601/dev/template/pyi-triage-select-smartphone/${language}`);
+      await page.click(".govuk-details__summary");
```
The docs mention something about using `.waitFor()` - do we need to do that?
>   // It is important to waitFor() the page to be in the desired
  // state *before* running analyze(). Otherwise, axe might not
  // find all the elements your test expects it to scan.

MikeCollingwood replied:
The tests passed without this but I will include just in case!

MikeCollingwood committed:
PYIC-7309:
- Use WEBSITE_HOST environment variable
- Add locator waitFor for open details test

MikeCollingwood committed:
Merge branch 'main' into pyic-7309

MikeCollingwood committed:
PYIC-7309: Update pagesAndContexts based on merge conflict

MikeCollingwood committed:
PYIC-7309: Remove unwanted file and update readme.md

MikeCollingwood committed:
PYIC-7309: Revert use of environment veriables in api-config for imposter

MikeCollingwood committed:
PYIC-7309: Update comment in accessibility.spec.js

blakeyp commented on browser-tests/functional-tests/accessibility.spec.js:
```diff
@@ -0,0 +1,45 @@
+        // Open all summaries on the page
+        const allSummaries = await page.locator("summary").all();
+        for (const summary of allSummaries) {
+          await summary.click();
+        }
+        const allSummaryDetails = await page.locator(".govuk-details__text").all();
+        for (const details of allSummaryDetails) {
+          await details.waitFor();
```
What's the difference between `click()` and `waitFor()` ?

MikeCollingwood replied:
The click is the action but in the second case we're waiting for the summaries to have opened, just to be sure they have

blakeyp replied:
Ah I see, thanks
